src/Product.java

```diff
@@ -0,0 +1,105 @@
+import java.time.LocalDate;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class Product {
+    private static long currentId;
+    private String type;
+    private double price;
+    private boolean discount;
+    private LocalDate date;
+    private long id;
+
+
+    public Product(String type, double price, boolean discount) {
+        this.type = type;
+        this.price = price;
+        this.discount = discount;
+        this.date = LocalDate.now();
+        this.id = currentId++;
+   }
+
+    public Product(String type, double price, boolean discount, LocalDate date) {
+        this.type = type;
+        this.price = price;
+        this.discount = discount;
+        this.date = date;
+        this.id = currentId++;
```

Comment:
You don't need to set your id here! It should be done via a setter somewhere in the business logic or pass it to the constructor (if passing through a constructor is an option for you - implement a builder pattern for this class, because 5 params are too many for 1 constructor).